imxrt-hal/src/can/mod.rs

```diff
@@ -0,0 +1,992 @@
+        can.begin();
+        can
+    }
+
+    pub fn instance(&mut self) -> &mut ral::can::Instance {
+        &mut self.reg
+    }
+
+    pub fn print_registers(&self) {
+        log::info!("MCR: {:X}", ral::read_reg!(ral::can, self.reg, MCR));
+        log::info!("CTRL1: {:X}", ral::read_reg!(ral::can, self.reg, CTRL1));
+        log::info!("CTRL2: {:X}", ral::read_reg!(ral::can, self.reg, CTRL2));
+        log::info!(
+            "RXMGMASK: {:X}",
+            ral::read_reg!(ral::can, self.reg, RXMGMASK)
+        );
+        log::info!(
+            "RXFGMASK: {:X}",
+            ral::read_reg!(ral::can, self.reg, RXFGMASK)
+        );
+
+        let max_fifo_filters = (read_reg!(ral::can, self.reg, CTRL2, RFFN) + 1) * 8;
+
+        for mailbox_number in 0..max_fifo_filters {
+            let mailbox_idflt_tab_addr =
+                self.mailbox_number_to_idflt_tab_address(mailbox_number as u8);
+            let idflt_tab =
+                unsafe { core::ptr::read_volatile((mailbox_idflt_tab_addr) as *mut u32) };
+            log::info!(
+                "IDFLT_TAB[{}, {:X}]: {:X}",
+                mailbox_number,
+                mailbox_idflt_tab_addr,
+                idflt_tab
+            );
+        }
+    }
+
+    fn while_frozen<F: FnMut(&mut Self) -> R, R>(&mut self, mut act: F) -> R {
+        let frz_flag_negate = ral::read_reg!(ral::can, self.reg, MCR, FRZACK == FRZACK_0);
+        self.enter_freeze_mode();
+        let res = act(self);
+        if frz_flag_negate {
+            self.exit_freeze_mode();
+        }
+        res
+    }
+
+    pub fn begin(&mut self) {
+        self.set_ccm_ccg();
+
+        ral::modify_reg!(ral::can, self.reg, MCR, MDIS: MDIS_0);
+
+        self.enter_freeze_mode();
+
+        ral::modify_reg!(ral::can, self.reg, CTRL1, LOM: LOM_1);
+        ral::modify_reg!(ral::can, self.reg, MCR, FRZ: FRZ_1);
+        while ral::read_reg!(ral::can, self.reg, MCR, LPMACK == LPMACK_1) {}
+
+        self.soft_reset();
+
+        while ral::read_reg!(ral::can, self.reg, MCR, FRZACK != FRZACK_1) {}
+
+        ral::modify_reg!(
+            ral::can,
+            self.reg,
+            MCR,
+            SRXDIS: SRXDIS_1,
+            IRMQ: IRMQ_1,
+            AEN: AEN_1,
+            LPRIOEN: LPRIOEN_1,
+            SLFWAK: SLFWAK_1,
+            WAKSRC: WAKSRC_1,
+            WRNEN: WRNEN_1,
+            WAKMSK: WAKMSK_1
+        );
+
+        ral::modify_reg!(ral::can, self.reg, MCR, |reg| reg & !0x8800);
+
+        ral::modify_reg!(
+            ral::can,
+            self.reg,
+            CTRL2,
+            RRS: RRS_1,
+            EACEN: EACEN_1,
+            MRP: MRP_1
+        );
+
+        self.disable_fifo();
+        self.exit_freeze_mode();
+    }
+
+    pub fn instance_number(&self) -> usize {
+        M::USIZE
+    }
+
+    pub fn is_can1(&self) -> bool {
+        M::USIZE == 1
+    }
+
+    pub fn is_can2(&self) -> bool {
+        M::USIZE == 2
+    }
+
+    pub fn base_address(&self) -> u32 {
```

Comment:
Could we make this private or `pub(crate)`? Or, why might the user need access to the base address?

Reply:
Yes, `pub(crate)` works here. No access to the base address should be required by the user.